redis/client.py

```diff
@@ -1290,8 +1294,21 @@ def execute_command(self, *args):
             self.connection.register_connect_callback(self.on_connect)
         connection = self.connection
         kwargs = {'check_health': not self.subscribed}
+        if not self.subscribed:
+            self.clean_health_check_responses()
         self._execute(connection, connection.send_command, *args, **kwargs)
 
+    def clean_health_check_responses(self):
+        """
+        If any health check responses are present, clean them
+        """
+        conn = self.connection
+        while self._execute(conn, conn.can_read, timeout=0):
+            response = self._execute(conn, conn.read_response)
+            if not self.is_health_check_response(response):
+                raise PubSubError('A non health check response was cleaned by '
+                                  'execute_command: {0}'.format(response))
```

Comment:
The idea of this function worries me. It depends on the response arriving in the socket before the function is called. What if the connection is suffering from high latency and the health check response only arrives after this function returns?

What about keeping a counter of outstanding health checks? Increment it when issuing a health check, and decrement it when a response is received (some care might be required to distinguish the two types of health check). Then you will know how many responses need to be cleared out.

Reply:
@bmerry 
I added a health check response counter. I also added a TTL so we won't end up stuck in the cleanup loop if the health check response isn't received for some reason. 
I setup the ttl to 10 seconds (the timeout of each can_read() is 1 second*10 times).
What do you think of this approach?